platform-sdk/swirlds-common/src/main/java/com/swirlds/common/config/sources/MappedConfigSource.java

```diff
@@ -0,0 +1,135 @@
+ * at runtime.
+ * </p>
+ * <p>
+ * Note that adding a mapping to a property not defined in the original {@link ConfigSource} will throw an
+ * {@link IllegalArgumentException} at runtime.
+ * </p>
+ * <p>
+ * Note that the ordinal of this {@code ConfigSource} is taken from the original {@link ConfigSource} that was wrapped
+ * by this class.
+ * </p>
+ *
+ * @see ConfigSource
+ * @see ConfigMapping
+ */
+public class MappedConfigSource extends AbstractConfigSource {
+
+    private static final Logger logger = LogManager.getLogger(MappedConfigSource.class);
+
+    private final ConfigSource wrappedSource;
+
+    private final Queue<ConfigMapping> configMappings;
+    private Map<String, String> properties;
+
+    /**
+     * Constructor that takes the wrapped config.
+     *
+     * @param wrappedSource the wrapped config
+     */
+    public MappedConfigSource(@NonNull final ConfigSource wrappedSource) {
+        this.wrappedSource = ArgumentUtils.throwArgNull(wrappedSource, "wrappedSource");
+        this.configMappings = new ConcurrentLinkedQueue<>();
+    }
+
+    /**
+     * Adds the mappedName {@code 'mappedName'<->'originalName'}
+     *
+     * @param mappedName   the mappedName name
+     * @param originalName the original name
+     */
+    public void addMapping(@NonNull final String mappedName, @NonNull final String originalName) {
+        addMapping(new ConfigMapping(mappedName, originalName));
+    }
+
+    /**
+     * Adds the mappedName {@code 'mappedName'<->'originalName'}
+     *
+     * @param configMapping defined mapping
+     */
+    public void addMapping(@NonNull final ConfigMapping configMapping) {
+        configMappings.add(configMapping);
+    }
+
+    @Override
+    @NonNull
+    protected Map<String, String> getInternalProperties() {
+        if (properties == null) {
+            final Map<String, String> internalProperties = wrappedSource.getProperties();
+            final Map<String, String> mappedProperties = new HashMap<>();
+            final Set<String> originals = new HashSet<>();
+
+            configMappings.forEach(configMapping -> {
+                if (internalProperties.containsKey(configMapping.mappedName())) {
+                    throw new IllegalArgumentException(
+                            "Property '" + configMapping.mappedName() + "' already found in original config source");
+                } else if (mappedProperties.containsKey(configMapping.mappedName())) {
+                    throw new IllegalArgumentException("Property '" + configMapping.mappedName() + "' already defined");
+                } else if (!internalProperties.containsKey(configMapping.originalName())) {
+                    throw new IllegalArgumentException(
+                            "Property '" + configMapping.originalName() + "' not found in original config source");
+                } else if (originals.contains(configMapping.originalName())) {
```

Comment:
Minor item.  I'm not sure we want this check.  I can easily see a case where we might want to migrate what is currently a single value to multiple values, and map all of the new keys to the original value for one release before adding the new keys (with possibly different values) in a later release.
Just something to think about.